### PR TITLE
In WorkSchemaGetEmbedding, replace 'type' property with 'config'

### PR DESCRIPTION
**Pull request by iguanaonmystack**
In WorkSchemaGetEmbedding, replace 'type' property with 'config', an abitrary object.

# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-445

#### Link to staging deployment URL 

https://ui-biomage-445.scp-staging.biomage.net/

#### Anything else the reviewers should know about the changes here

I'm not sure this is done but I wanted to make sure I haven't misunderstood the task. This would allow any object as defined by convention between the UI and worker as long as it has a name and 'config' member, which would contain the type, UI settings and so on. As mentioned in the ticket, making this any more rigidly defined would slow down faster iteration of both components.

# Changes
### Code changes

* Swagger yaml spec updated

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock
- [x] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [x] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/specs/api.yaml [hunk 1]
```diff
@@ -458,16 +458,11 @@ components:
...
-        type:
-          type: string
-          enum:
-            - pca
-            - tsne
-            - umap
-          description: The type of embedding to compute.
+        config:
+          type: object
+          description: The configuration settings to use for this embedding. This will be used both to configure the embedding and to cache the results.
```
Comment (ivababukova) [inline src/specs/api.yaml]: I would rephrase this description as `The configuration settings to use for calculation of this embedding.` I wouldn't have the `This will be used both to configure the embedding and to cache the results.` part because the caching applies to all the tasks, not only to the embedding

File: src/specs/api.yaml [hunk 2]
```diff
@@ -458,16 +458,11 @@ components:
           type: string
           example: GetEmbedding
           description: The name of this work -- must be `GetEmbedding`.
```
Comment (ivababukova) [inline src/specs/api.yaml:461]: I see that there is no schema defined for the work response for GetEmbedding -- can you add one too? This will make our job easier in the future when we decide to enable the validation of responses received from the worker.

**Comment by marcellp** [comment]
@ivababukova -- my understanding is that this will break the worker unless we merge the other changes in first.

@iguanaonmystack -- could you deploy this to staging to make sure things work as intended?

**Comment by iguanaonmystack** [comment]
@marcellp I've deployed it here: https://ui-98rxkh16615sjk74pbu3y85j9a.scp-staging.biomage.net/experiments/5e959f9c9f4b120771249001/data-exploration

**Comment by iguanaonmystack** [comment]
I wasn't sure which bit Marcel indicated might break, so I've added the WorkResponseGetEmbedding anyway. I'm not sure what tests it needs at this point if we're waiting on other changes first?

**Comment by ivababukova** [comment]
@marcellp I don't think it will break the worker, why would it? Is the worker using the `type` parameter in the `GetEmbedding` task at the moment?

**Comment by ivababukova** [comment]
I took a look into the worker code just now, Marcell is right and this will break the worker, see: https://github.com/biomage-ltd/worker/blob/master/python/src/tasks/embedding.py#L55 and also see the UI: https://github.com/biomage-ltd/ui/blob/master/src/redux/actions/embedding/loadEmbedding.js#L22 At the moment we are using `type`, so we have to keep using type. In order to be able to merge this now without breaking anything, you should undo the changes that remove "type" from the schema and only have `config` added as an additional optional parameter.

**Comment by iguanaonmystack** [comment]
I hadn't realised that it would be a problem if the existing code was sending an additional property the schema didn't recognise. I've added `type` back in now

**Comment by ivababukova** [comment]
The problem is not with the additional property, but with the deletion of the property`type`

**Comment by marcellp** [comment]
@ivababukova -- I thought it would be fine, but my concern was the `type` parameter in the worker, yes.

@iguanaonmystack shares my understanding, as far as I know OpenAPI schemas in general allow additional fields to be declared (such as `type` in this case) as long as they do not interfere with something that actually /is/ in the spec (i.e. there is a field `settings` that is declared to be an object in the schema, it *must* be an object, and not a string).

That said, I wasn't entirely sure, so I wanted to see it work in staging before merging.

**Review by marcellp — DISMISSED** [review]
I think @iguanaonmystack was correct, but the system is actually broken in staging. We have the following error:

```
[2021-01-22T11:06:06.673Z] Error while processing WorkRequest event: Error: Error: bodymatches more than one entry in a oneOf.
    at validateRequest (/app/src/utils/schema-validator.js:16:11)
    at handleWorkRequest (/app/src/api/event-services/work-request.js:28:7)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[2021-01-22T11:06:06.673Z] Trace: Error: Error: bodymatches more than one entry in a oneOf.
    at validateRequest (/app/src/utils/schema-validator.js:16:11)
    at handleWorkRequest (/app/src/api/event-services/work-request.js:28:7)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Object.trace (/app/src/utils/logging.js:15:13)
    at Socket.<anonymous> (/app/src/api/events/index.js:13:14)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

Seems like one of the new schemas is not well-described. Could you look into why this is happening?

File: src/specs/api.yaml [hunk 3]
```diff
@@ -458,16 +458,27 @@ components:
...
+          type: object
+          description: The configuration settings to use for calculation of this embedding.
         type:
           type: string
           enum:
             - pca
             - tsne
             - umap
-          description: The type of embedding to compute.
+          description: The type of embedding to compute (deprecated; use config).
```
Comment (marcellp) [inline src/specs/api.yaml]: You can add a deprecated tag for this in Stoplight Studio, it's a good idea because then we could check later for deprecated stuff being used.

File: src/specs/api.yaml [hunk 4]
```diff
@@ -458,16 +458,27 @@ components:
           type: string
           example: GetEmbedding
           description: The name of this work -- must be `GetEmbedding`.
```
Comment (marcellp) [inline src/specs/api.yaml]: My understanding is that the API is failing because you did not set up a pattern for the "type" string like it is done [here](https://github.com/biomage-ltd/api/blob/02d51c45bf6698128832c97558cc7b6986fe082c/src/specs/api.yaml#L489) -- for this schema, only `name` is required, and name does not have a pattern. Therefore, any and all work tasks will match with this, as well as with their respective request schemas.

**Comment by marcellp** [comment]
In general, I would strongly encourage using Stoplight Studio for this -- the OpenAPI schema is really complicated and we have always found things got messed up when we manually edited the schema.

**Review by ivababukova — DISMISSED** [review]

**Comment by ivababukova** [comment]
> I think @iguanaonmystack was correct, but the system is actually broken in staging. We have the following error:
> 
> ```
> [2021-01-22T11:06:06.673Z] Error while processing WorkRequest event: Error: Error: bodymatches more than one entry in a oneOf.
>     at validateRequest (/app/src/utils/schema-validator.js:16:11)
>     at handleWorkRequest (/app/src/api/event-services/work-request.js:28:7)
>     at runMicrotasks (<anonymous>)
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
> [2021-01-22T11:06:06.673Z] Trace: Error: Error: bodymatches more than one entry in a oneOf.
>     at validateRequest (/app/src/utils/schema-validator.js:16:11)
>     at handleWorkRequest (/app/src/api/event-services/work-request.js:28:7)
>     at runMicrotasks (<anonymous>)
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
>     at Object.trace (/app/src/utils/logging.js:15:13)
>     at Socket.<anonymous> (/app/src/api/events/index.js:13:14)
>     at runMicrotasks (<anonymous>)
>     at processTicksAndRejections (internal/process/task_queues.js:97:5)
> ```
> 
> Seems like one of the new schemas is not well-described. Could you look into why this is happening?

@iguanaonmystack how is that going? Is the cause of the error found and fixed?

**Comment by iguanaonmystack** [comment]
I'd given Stoplight a go but it removed the single quotes from a whole bunch of strings in the yaml, which made everything difficult to diff. Does this happen for anyone else; are the extra changes it puts in just generally ignored?

**Comment by iguanaonmystack** [comment]
@marcellp Where do I find those logs in staging? I've been poking around in Kontena Lens and the best I've managed is a shell on the node. I seem to recall when you demoed it you just clicked a button somewhere and got a terminal tailing the logs/stdout somehow?

**Comment by iguanaonmystack** [comment]
I've tried to re-deploy the staging site but I'm getting a permissions error: https://github.com/biomage-ltd/iac/actions/runs/509079936

**Comment by iguanaonmystack** [comment]
Removing the staging site and re-creating it (https://ui-biomage-445.scp-staging.biomage.net/experiments/5e959f9c9f4b120771249001/data-exploration) seems to have worked. This looks okay to me now?

**Review by ivababukova — APPROVED** [review]
